install.sh

```diff
@@ -0,0 +1,212 @@
+#!/bin/sh
+
+set -e
+
+DOTENV_LINTER="dotenv-linter"
+DOTENV_LINTER_REPO="${DOTENV_LINTER}/${DOTENV_LINTER}"
+DOTENV_LINTER_GITHUB="https://github.com/${DOTENV_LINTER_REPO}"
+DOTENV_LINTER_RELEASES="${DOTENV_LINTER_GITHUB}/releases"
+DOTENV_LINTER_RELEASE_API="https://api.github.com/repos/${DOTENV_LINTER_REPO}/releases/latest"
+
+usage() {
+    require_cmd cat
+    this=$1
+    cat 1>&2 <<EOF
+$this: download binaries for dotenv-linter
+
+USAGE:
+    $this [FLAGS] [OPTIONS] <tag>
+
+FLAGS:
+    -h, --help      Prints help information
+
+OPTIONS:
+    -b, --bindir <DIR_PATH>     Sets bindir or installation directory. Defaults to ./bin
+
+ARGS:
+    <tag>       is a tag from ${DOTENV_LINTER_RELEASES}. If tag is missing, then the latest will be used.
+EOF
+    exit 2
+}
+
+parse_args() {
+  BINDIR=${BINDIR:-./bin}
+  while [ "$#" -gt 0 ]; do
+    case $1 in
+        -h|--help)
+            usage "$0"
+            shift # past argument
+            ;;
+        -b|--bindir)
+            BINDIR="$2"
+            shift # past argument
+            shift # past value
+            ;;
+        *) TAG=$1
+            shift # past argument
+            ;;
+    esac
+  done
+}
+
+main() {
+    parse_args "$@"
+
+    require_cmd uname
+    require_cmd mktemp
+    require_cmd grep
+    require_cmd rm
+
+    get_architecture || return 1
+    _arch="$RETVAL"
+
+    case $_arch in
+        win-*) _ext=".zip" ;;
+        *) _ext=".tar.gz" ;;
+    esac
+
+    _archive_name="${DOTENV_LINTER}-${_arch}${_ext}"
+
+    if [ -z "${TAG}" ]; then
+        get_latest_version || return 1
+        _target_version="$RETVAL"
+    else
+        _target_version="${TAG}"
+    fi
+
+    println "Version ${_target_version} will be installed"
+    _url="${DOTENV_LINTER_RELEASES}/download/${_target_version}/${_archive_name}"
+
+    # Installation
+    _temp_dir=$(mktemp -d)
+    _archive_path="${_temp_dir}/${_archive_name}"
+
+    download "${_url}" "${_archive_path}" || return 1
+
+    case $_archive_path in
+        *.zip)
+            required_cmd unzip
+            unzip "${_archive_path}" -d "${_temp_dir}"
+            _file_name="${DOTENV_LINTER}.exe"
+        ;;
+        *)  require_cmd tar
+            tar -xzf "${_archive_path}" -C "${_temp_dir}"
+            _file_name="${DOTENV_LINTER}"
+        ;;
+    esac
+
+    test ! -d "${BINDIR}" && install -d "${BINDIR}"
+
+    install "${_temp_dir}/${_file_name}" "${BINDIR}/"
+    println "Successfully installed to ${BINDIR}/${_file_name}"
+
+    rm -rf "${_temp_dir}"
+
+    return 0;
+}
+
+get_latest_version() {
+    require_cmd cut
+
+    download "${DOTENV_LINTER_RELEASE_API}" || return 1
+    _latest_version=$(
+         echo "$RETVAL" | grep '"tag_name":' | cut -d'"' -f4
+    )
+
+    RETVAL=${_latest_version}
```

Comment:
It seems like `RETVAL` is a global variable, I've never seen this pattern until now but it seems usable.
Personally I like the `echo` way a bit better, as it eliminates the usage of global variables.

You can read about it here in example 2: https://linuxhint.com/return-string-bash-functions/

Anyway - if you want to stick to the `RETVAL` way, go for it. :)

Reply:
Actually, I liked this approach in sh scripts.
I met him a couple of times, for example in a [rustup instillation script](https://sh.rustup.rs/) :relaxed:

Reply:
Okay, let's leave it like this :) It is just a personal opinion and I don't think this will be the reason why our install script will break at some point.